promql/engine.go

```diff
@@ -743,9 +747,14 @@ func (ev *evaluator) rangeEval(f func([]Value, *EvalNodeHelper) Vector, exprs ..
 					Points: getPointSlice(numSteps),
 				}
 			}
-			sample.Point.T = ts
-			ss.Points = append(ss.Points, sample.Point)
-			seriess[h] = ss
+			if curSamples < maxSamples {
```

Comment:
it seems in some cases that there's the potential for _no_ samples to be added in the above block of for loops, and so they're then added later down here, so we have to track curSamples here as well

Reply:
Okay, so the output is the correct place to do it then? Or should we limit the input as well? The goal is to reduce the chance of OOM by limit the # of samples returned by a query, in this case are we not doubling up on memory allocation with the `vectors` for the input to the functions `f`, and then building the output?

Reply:
You could view it a few ways. It probably doesn't hurt to limit both.

Reply:
So in some cases, the input (or rather, output of `f`) is the only input that we consider when building the output vector. We don't want every function passed to `rangeEval` to have to deal with limiting the # of samples right?

Reply:
Yes, only matrix functions and raw lookups should need special handling beyond that